dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,239 @@
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.cached_dirs = {}
+        self.cached_ids = {}
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        if Config.SECTION_REMOTE_KEY_FILE not in self.config:
+            raise DvcException(
+                "Google Drive settings file path is missed from config. "
+                "Learn more at https://dvc.org/doc."
+            )
+        self.gdrive_credentials_path = self.config.get(
+            Config.SECTION_REMOTE_KEY_FILE
+        )
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    @sleep_and_retry
+    @limits(calls=10, period=1)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            if "Rate Limit Exceeded" in str(exception):
+                raise DvcException("API usage rate limit exceeded")
+            raise
+        return result
+
+    def list_drive_item(self, query):
+        list_request = RequestListFilePaginated(self.drive, query)
+        page_list = self.execute_request(list_request)
+        while page_list:
+            for item in page_list:
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            self.cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            self.cached_ids[dir1["id"]] = dir1["title"]
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv("PYDRIVE_USER_CREDENTIALS_DATA"):
+            with open("credentials.json", "w") as credentials_file:
+                credentials_file.write(
+                    os.getenv("PYDRIVE_USER_CREDENTIALS_DATA")
+                )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        gauth = GoogleAuth(settings_file=self.gdrive_credentials_path)
+        gauth.CommandLineAuth()
+        gdrive = GoogleDrive(gauth)
+        return gdrive
+
+    def create_drive_item(self, parent_id, title):
+        upload_request = RequestUploadFile(
+            {
+                "drive": self.drive,
+                "title": title,
+                "parent_id": parent_id,
+                "mime_type": FOLDER_MIME_TYPE,
+            }
+        )
+        result = self.execute_request(upload_request)
+        return result
+
+    def get_drive_item(self, name, parents_ids):
+        query = " or ".join(
+            "'{}' in parents".format(parent_id) for parent_id in parents_ids
+        )
+        if not query:
```

Comment:
better check parent_ids before `query = <something>` if we are sure that this can happen only in that case, otherwise someone can change the previous assignment and we will be returning `None` in some strange situations